### PR TITLE
Add syntax for mysql specific SELECT statements

### PR DESCRIPTION
**Pull request by silverbullettruck2001**
Shoutout to @barrywhart, @dmateusp and @NiallRees for their help and responsiveness to my questions regarding this enhancement!  Thanks again!

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_mysql.py [hunk 1]
```diff
@@ -53,11 +77,29 @@
             Ref("DoubleQuotedLiteralSegment"),
         ]
     ),
-    PostTableExpressionGrammar=Sequence(
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_mysql.py:56]: Why was this removed?

File: src/sqlfluff/dialects/dialect_mysql.py [hunk 2]
```diff
@@ -53,11 +77,29 @@
...
-        Bracketed(Ref("ObjectReferenceSegment")),
+    FromClauseTerminatorGrammar=OneOf(
+        "WHERE",
+        "LIMIT",
+        "GROUP",
+        "ORDER",
+        "HAVING",
+        "QUALIFY",
+        "WINDOW",
+        Ref("IndexHintClauseSegment"),
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_mysql.py:88]: It looks like this element is the same as the ANSI dialect except for lines 88-90. Should we look at using `.copy()` to base this segment on the ANSI one, to reduce copy/paste? (NOTE: I _think_ this is supported for grammars as well as segments.)

File: src/sqlfluff/dialects/dialect_mysql.py [hunk 3]
```diff
@@ -53,11 +77,29 @@
...
+        Ref("WithNoSchemaBindingClauseSegment"),
+    ),
+    BaseExpressionElementGrammar=OneOf(
+        Ref("LiteralGrammar"),
+        Ref("BareFunctionSegment"),
+        Ref("FunctionSegment"),
+        Ref("IntervalExpressionSegment"),
+        Ref("ColumnReferenceSegment"),
+        Ref("ExpressionSegment"),
+        Ref("SessionVariableNameSegment"),
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_mysql.py:101]: Similar to the above, it looks like this is the same as ANSI except for these two added items. Consider using `.copy()` here as well?

File: src/sqlfluff/dialects/dialect_mysql.py [hunk 4]
```diff
@@ -475,6 +517,214 @@ class DefinerSegment(BaseSegment):
...
+            OneOf("NOWAIT", Sequence("SKIP", "LOCKED"), optional=True),
+        ),
+        Sequence("LOCK", "IN", "SHARE", "MODE"),
+        optional=True,
+    )
+
+
+@mysql_dialect.segment()
+class IndexHintClauseSegment(BaseSegment):
+    """This is the body of a index hint clause."""
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_mysql.py:710]:
```suggestion
    """This is the body of an index hint clause."""
```

File: test/fixtures/parser/mysql/call_statement.sql [hunk 5]
```diff
@@ -1 +1 @@
-CALL zproc_validate_send_adminuser_otp('a');
\ No newline at end of file
+CALL somefunction('a');
```
Comment (barrywhart) [inline test/fixtures/parser/mysql/call_statement.sql:1]: Is there a reason for changing the function name in this and other test `.sql` files?

File: test/fixtures/parser/mysql/force_index.sql [hunk 6]
```diff
@@ -1,8 +1 @@
-SELECT
-       onetable.f1,
-       twotable.f1
-FROM onetable
-         LEFT JOIN
-     twotable
-         FORCE INDEX FOR JOIN (idx_index)
-     ON onetable.f1 = twotable.f1
\ No newline at end of file
+SELECT * FROM onetable FORCE INDEX (idx_index);
```
Comment (barrywhart) [inline test/fixtures/parser/mysql/force_index.sql:1]: Why so many changes to this file? Could this (should this) be a new test file instead?

**Review by barrywhart — CHANGES_REQUESTED** [review]
This looks pretty close to ready! I have a few questions and suggestions.

File: src/sqlfluff/dialects/dialect_mysql.py (hunk 1 above)
Comment (silverbullettruck2001) [reply src/sqlfluff/dialects/dialect_mysql.py:56]: > Why was this removed?

It wasn't needed after I expanded the scope for the INDEX functionality by added in the IndexHintClauseSegment at the bottom of the dialect definition so I removed this.

File: test/fixtures/parser/mysql/call_statement.sql (hunk 5 above)
Comment (silverbullettruck2001) [reply test/fixtures/parser/mysql/call_statement.sql:1]: I used a name of a proc that is specific to me and wanted to make it generic instead.

File: test/fixtures/parser/mysql/force_index.sql (hunk 6 above)
Comment (silverbullettruck2001) [reply test/fixtures/parser/mysql/force_index.sql:1]: I simplified this one since the original statement for the test was more complex than necessary.  I will add a new test that is specific to using the `FORCE INDEX` on a join so that the original test will still be validated.

File: src/sqlfluff/dialects/dialect_mysql.py (hunk 4 above)
Comment (silverbullettruck2001) [reply src/sqlfluff/dialects/dialect_mysql.py:710]: > 

I will get this correction made

File: test/fixtures/parser/mysql/force_index.sql (hunk 6 above)
Comment (silverbullettruck2001) [reply test/fixtures/parser/mysql/force_index.sql:1]: @barrywhart good catch, after adding the test back, my changes no longer covered it.  But I have it updated now to cover both this test and my existing tests.  Thanks!

File: src/sqlfluff/dialects/dialect_mysql.py (hunk 3 above)
Comment (silverbullettruck2001) [reply src/sqlfluff/dialects/dialect_mysql.py:101]: Will do, good suggestion.

File: src/sqlfluff/dialects/dialect_mysql.py (hunk 2 above)
Comment (silverbullettruck2001) [reply src/sqlfluff/dialects/dialect_mysql.py:88]: Will do, good suggestion.

**Comment by silverbullettruck2001** [comment]
> This looks pretty close to ready! I have a few questions and suggestions.

@barrywhart I made the suggested changes, let me know if there's anything else that needs to be adjusted.

**Review by barrywhart — APPROVED** [review]

**Comment by NiallRees** [comment]
Your commitment to this has been great to see @silverbullettruck2001!

**Comment by barrywhart** [comment]
PR looks good! Will merge once the build passes.

**Comment by silverbullettruck2001** [comment]
> Your commitment to this has been great to see @silverbullettruck2001!

Thanks @NiallRees!
